app/scripts/metamask-controller.js

```diff
@@ -185,10 +184,25 @@ module.exports = class MetamaskController {
       },
       // tx signing
       approveTransaction: this.newUnsignedTransaction.bind(this),
-      signTransaction: (...args) => {
-        this.setupSigningListners(...args)
-        this.txManager.formatTxForSigining(...args)
-        this.sendUpdate()
+      signTransaction: (txParams, cb) => {
+        this.txManager.formatTxForSigining(txParams)
+        .then(({ethTx, address, txId}) => {
+          return this.keyringController.signTransaction(ethTx, address, txId)
+        })
+        .then(({tx, txId}) => {
+          return this.txManager.resolveSignedTransaction({tx, txId})
```

Comment:
Fun alternative writing of these two lines, since the outputs match the inputs:
```
.then(this.txManager.resolveSignedTransaction.bind(this.txManager))
```
More readable if you already had a local txManager reference:
```
.then(txManager.resolveSignedTransaction.bind(txManager))
```
Another one-liner:
```
.then(result => txManager.resolveSignedTransaction(result))
```